libOTe/Vole/Subfield/NoisyVoleReceiver.h

```diff
@@ -0,0 +1,104 @@
+// © 2022 Visa.
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// This code implements features described in [Silver: Silent VOLE and Oblivious
+// Transfer from Hardness of Decoding Structured LDPC Codes,
+// https://eprint.iacr.org/2021/1150]; the paper is licensed under Creative
+// Commons Attribution 4.0 International Public License
+// (https://creativecommons.org/licenses/by/4.0/legalcode).
+
+#include <libOTe/config.h>
+#if defined(ENABLE_SILENT_VOLE) || defined(ENABLE_SILENTOT)
+
+#include "cryptoTools/Common/Defines.h"
+#include "cryptoTools/Common/Timer.h"
+#include "cryptoTools/Crypto/PRNG.h"
+#include "libOTe/Tools/Coproto.h"
+#include "libOTe/TwoChooseOne/OTExtInterface.h"
+
+namespace osuCrypto {
+
+template <typename TypeTrait>
+class NoisySubfieldVoleReceiver : public TimerAdapter {
+ public:
+  using F = typename TypeTrait::F;
+  using G = typename TypeTrait::G;
+  task<> receive(span<G> y, span<F> z, PRNG& prng,
+                                    OtSender& ot, Socket& chl) {
+    MC_BEGIN(task<>, this, y, z, &prng, &ot, &chl,
+             otMsg = AlignedUnVector<std::array<block, 2>>{sizeof(F) * 8}); // todo: sizeof(F) * 8
+
+    setTimePoint("NoisyVoleReceiver.ot.begin");
+
+    MC_AWAIT(ot.send(otMsg, prng, chl));
+
+    setTimePoint("NoisyVoleReceiver.ot.end");
+
+    MC_AWAIT(receive(y, z, prng, otMsg, chl));
+
+    MC_END();
+  }
+
+  task<> receive(span<G> y, span<F> z, PRNG& _,
+                                    span<std::array<block, 2>> otMsg,
+                                    Socket& chl) {
+    MC_BEGIN(task<>, this, y, z, otMsg, &chl, msg = Matrix<F>{},
+             prng = std::move(PRNG{})
+    );
+
+    setTimePoint("NoisyVoleReceiver.begin");
+    if (otMsg.size() != sizeof(F) * 8) throw RTE_LOC;
+    if (y.size() != z.size()) throw RTE_LOC;
+    if (z.size() == 0) throw RTE_LOC;
+
+    memset(z.data(), 0, sizeof(F) * z.size());
+    msg.resize(otMsg.size(), z.size());
```

Comment:
When F=G^m with large m, this would eat all memory, given `z.size()` is already 400 from silent VOLE. We don't need a matrix and I think it's here for chl performance? Any idea to improve the performance?

Reply:
Each element of `msg` is in F, and `otMsg` has the bit length of `F`, so this communication(and memory now) is m^2 complexity.

Reply:
Suppose G has k bits, F=G^m, and length of the noisy subfield VOLE (numPartitions) be `n`. Currently, for bit of the delta in F, the OT message has `(2*) n` elements in F, which is `(2*) n*k*m` bits each. Since F has `k*m` bits, the overall communication complexity is `n*k^2*m^2`.

Alternatively, we do `m` times noisy VOLE with F=G. Each time the communication complexity is `O(n*k^2)`, so overall is `O(n*m*k^2)`. I assume the security is the same. We should be able to use template to check if F!=G(which means F=G^m for m > 1), and use this approach.

What do you think?

Reply:
**Current Design:** We want to compute `a * X + B =  C` for `a in G^n`, `B,C in F^n` and `X in F`, i.e. `X=delta`. Let `k = log2(|G|), t=km = log2(|F|)`. 

The number of OTs is `t`. not `n`. We do an OT for each bit of `X=delta`.

The overhead of the protocol is sending `t` elements of `F^n` That is, for each of the `t` bits of `X`, we do an OT, and then send a single vector in `F^n`. Therefore `t^2 * n = k^2 * m^2 *n` bits in total.

**Alternative Design:** There is an alternative design where you do an OT for each bit of `a`. You would then need `kn` OTs. This will be a lot more in most cases.

Are you suggesting this?

**Multi-threading ?:** You could but we we only do very small noisyVoles, so I dont think it makes sense to make the noisy vole itself parallel. I'd think if you wanted to do that, you could just have the caller make multiple noisy voles and call those in parallel.

Reply:
No, I am suggesting doing `m` times of noisy G-G VOLE, each VOLE has `k` OT messages, so it's still `m*k` messages. But since OT messages are sending elements of G now, not F, so message size for each OT message is reduced from `n*k*m` to `n*m`.

Reply:
I agree multi-threading would not make a huge improvement. I just wanted to say it's possible.

Reply:
ah, right, so when we compute 
```
vi = a * F(2^i)
```
as described above, you could naively represent this as a `F^n` vector but you're saying that's kinda wasteful because you can also just represent it as 
```
vi = a  G(2^{i mod k}) 
```
The sender can then scale these up on their side by multiplying by `F(2^{i - (i mod k)})`. Yeah, makes sense. nice.

Reply:
Yeah, I'm going to implement this.

Reply:
I guess you could also change how this works in that you dont need "pre" multiply with `2^i` at all. Instead, you could have
```
vi = a
```
do OT on 
```
mi0 = Ri
mi1 = Ri + vi = Ri + a
```
and then have both parties scale `mi0,mi1` up by `F(2^i)`

Reply:
This means both end need to do a multiplication, but know only the receiver need to do a multiplication.

Also, I'm not sure about the security of my approach mentioned above. We need the receiver to input the same y for `m` times. I think it should be secure though.

Reply:
I guess in both cases you need both parties to do the scaling. Either way is fine.

Security should be fine. You only need Semi-honest security and privacy against malicious parties. We don't require correctness. So changing the `a=y` is fine.